src/realm/alloc.cpp

```diff
@@ -116,4 +120,62 @@ Allocator& Allocator::get_default() noexcept
 {
     return default_alloc;
 }
+
+// This function is called to handle translation of a ref which is above the limit for its
+// memory mapping. This requires one of three:
+// * bumping the limit of the mapping. (if the entire array is inside the mapping)
+// * adding a cross-over mapping. (if the array crosses a mapping boundary)
+// * using an already established cross-over mapping. (ditto)
+// this can proceed concurrently with other calls to translate()
+char* Allocator::x_translate_extended(RefTranslation* ref_translation_ptr, ref_type ref) const noexcept
+{
+    size_t idx = get_section_index(ref);
+    RefTranslation& txl = ref_translation_ptr[idx];
+    size_t offset = ref - get_section_base(idx);
+    char* addr = txl.mapping_addr + offset;
+#if REALM_ENABLE_ENCRYPTION
+    realm::util::encryption_read_barrier(addr, NodeHeader::header_size, txl.encrypted_mapping, nullptr);
+#endif
+    auto size = NodeHeader::get_byte_size_from_header(addr);
+    bool crosses_mapping = offset + size > (1 << section_shift);
+    if (REALM_LIKELY(!crosses_mapping)) {
+        // Array fits inside primary mapping, no new mapping needed, just move the limit on
+        // use of the existing primary mapping.
+        // Take into account that another thread may attempt to change / have changed it concurrently,
+        // by re-evaluating whether we still need to bump. In case of conflict we back off.
+        size_t mapping_limit = txl.lowest_possible_xover_offset.load(std::memory_order_relaxed);
+        if (offset + size > mapping_limit) {
+            txl.lowest_possible_xover_offset.compare_exchange_weak(mapping_limit, offset + size,
```

Comment:
On second thought, I'd hoist this up outside of the `if (!crosses_mapping)` block and move it to the top of the function. Then remove the store in `SlabAlloc::get_or_add_xover_mapping` since it becomes redundant.

Rationale: If we see a ref at offset X, then we know that any ref at offset < X must not be xover because it must end before X because otherwise it would contain X. This holds regardless of whether or not X is itself an xover array. Therefore we might as well update the `lowest_possible_xover_offset` ASAP rather than waiting until we've done a ton of bookkeeping, possibly even going to disk. This will further reduce the time that other threads would need to enter the slow path for refs that are not possibly xover.

Reply:
Really good!

Reply:
On second thought: bad!

We can't do that, because the limit is not supposed to move in the case where we actually create a new mapping. Only when we don't.

Reply:
Isn't this bumping the limit when you create the mapping? https://github.com/realm/realm-core/blob/e85a052b0893c13ffd348d691b5c489b11f201b7/src/realm/alloc_slab.cpp#L1334-L1335

Reply:
no, it only stores the offset, not the offset + size. Later attempts to translate the same offset needs to be "caught" and take the path into this function. Always incrementing the limit leads to 
 us filtering later attempts to translate the same reference too early.

Reply:
Oh, of course. I forgot that the !xover path was setting it to the start of the next array while the xover path set it to the start of this array. We *could* do something like bump it to the start of this array, then if it isn't xover bump it again to the next array, but I can't tell if that would actually be better. It would definitely be slower for single-threaded cases, so it is probably a bad idea.

Could you hoist the `lowest_possible_xover_offset.store()` in the xover case linked above to be sooner, in particularly before grabbing the lock? Then resolve this thread.